Add why meditate page

alshakerM opened the pull request.

alshakerM committed:
Add why meditate page

alshakerM committed:
Remove margin from lessons

alshakerM committed:
Add audio player with subs

alshakero commented on components/WhyMeditate/WhyMeditate.js:
```diff
@@ -0,0 +1,59 @@
+export function WhyMeditate() {
+  const [pause, setPause] = React.useState(false);
+  const [currentProgress, setCurrentProgress] = React.useState(0);
+  console.log(currentProgress);
```
```suggestion

```

alshakero commented on components/WhyMeditate/AudioWithSubs.js:
```diff
@@ -0,0 +1,57 @@
+            </span>
+          ))}
+        </p>
+      ))}
```
This sometimes has one segment, and at other times it can have two. This makes the pause button jump up and down based on the number of segments. Can you come up with a way to make sure this segment always has the same height?

alshakero commented on components/WhyMeditate/WhyMeditate.js:
```diff
@@ -0,0 +1,59 @@
+        concise case for mediation in this recording.
+      </p>
+      <div
+        className={cx(styles.playerContainer, { [styles.isPlayed]: !pause })}
```
`{ [styles.isPlayed]: !pause }` this is a bit of a problem, because the default state is paused. 

See how the default state looks,  even though it's not playing

![image](https://user-images.githubusercontent.com/17054134/150796502-600f5e0f-d65b-48eb-8b46-75c9bda4c093.png)

alshakero commented on components/WhyMeditate/WhyMeditate.js:
```diff
@@ -0,0 +1,59 @@
+        <div
+          className={cx(styles.audioContainer, { [styles.isPlayed]: !pause })}
+        >
+          {!pause && (
+            <>
+              <AudioWithSubs
+                pause={pause}
+                setCurrentProgress={setCurrentProgress}
+              />
+            </>
+          )}
```
Because of line 37, you're not rendering the AudioWithSubs at all when paused. Meaning you're deleting it and its progress. That's why you're losing progress when you pause.

alshakero commented on public/subs.json:
```diff
@@ -0,0 +1,2838 @@
+[
```
Since you're importing and not fetching, this file can stay inside the component's folder, doesn't need to be in public.

alshakero reviewed (changes requested):
Looks nice, but there is a couple bugs.

alshakerM committed:
Address feedback

alshakerM commented:
Deleted console log.
Added a state to monitor the player size.
Changed the background of the active words in subs,
the stroke width of the progress from 5 to 4,
and play icon size when pause in big sized player.
Moved subs to WhyMeditate component.

alshakero commented on components/WhyMeditate/AudioWithSubs.module.css:
```diff
@@ -0,0 +1,33 @@
+  margin-top: 10px;
+}
+.audioContainer {
+  min-height: 109px;
```
Such a brilliant, simple solution.

alshakero commented on components/WhyMeditate/AudioWithSubs.module.css:
```diff
@@ -0,0 +1,33 @@
+
+@keyframes appear {
+  0% {
+    background: #a85a5d;
```
I think if you change the color instead of the background, it would look even nicer.

alshakero commented on components/WhyMeditate/WhyMeditate.js:
```diff
@@ -0,0 +1,66 @@
+            <>
+              <AudioWithSubs
+                pause={pause}
+                setCurrentProgress={setCurrentProgress}
```
```suggestion
      	setCurrentProgress={(progress) => {
          setCurrentProgress(progress);
          if (progress === 1) {
            setIsPlayed(false);
          }
        }}
```

alshakero reviewed (approved):
Looks soooooo beautiful, I love it sooooooo much.

💡 **Optional**: For artistic value, I'd create a list of special words and show them in italic or a special color, so they'd stand out.

Something like: 

```js
const specialWords = ['reality', 'confuse', 'meditation', 'world', 'symbols', 'minds', 'wealth', 'images', 'dangerous', 'myself', 'think', 'madness', 'compulsively'];

const isSpecial = specialWords.includes(word.trim());

className={cx(styles.word, {isSpecial})}
```

alshakerM committed:
Add specialWords to style to subs

alshakerM commented:
Added special words and styled them with animation, and a playback rate button.
Changed the code to SVG instead of a div for mobile support, and play pause icon colors.

alshakero commented on components/WhyMeditate/AudioWithSubs.js:
```diff
@@ -0,0 +1,76 @@
+
+  return findSegments(time * 1000).map((segment, index) => {
+    const allWords = segment.parts.reduce((words, word) => (words += word), '');
+    const specialWord = segment.parts.find((word) =>
+      specialWords.includes(word.trim())
+    );
```
I'm not a big fun of this effect anymore. Let's get rid of big special words and only keep the smaller ones.

alshakero commented on components/WhyMeditate/WhyMeditate.module.css:
```diff
@@ -0,0 +1,110 @@
+  animation: appear 100ms ease-in forwards;
+  opacity: 0;
+}
+.speedButton {
+  border-radius: 50%;
+  border: none;
+  min-height: 50px;
+  min-width: 50px;
+  font-size: 1.5em;
+  font-style: italic;
+  color: #a85a5d;
+  animation: delay 1.3s ease-out;
+}
```
I think it would look nicer like this: 

```suggestion
.speedButton {
   top: -80px;
    border-radius: 50%;
    border: none;
    min-height: 50px;
    background: none;
    min-width: 50px;
    font-size: 1.2em;
    font-style: italic;
    color: #a85a5d;
    position: relative;
}
```

alshakero commented on components/WhyMeditate/WhyMeditate.module.css:
```diff
@@ -0,0 +1,110 @@
+  animation: delay 1.3s ease-out;
+}
+@keyframes delay {
+  0% {
+    opacity: 0;
+  }
+  25% {
+    opacity: 0.25;
+  }
+  50% {
+    opacity: 0.5;
+  }
+  76% {
+    opacity: 0.75;
+  }
+  100% {
+    opacity: 1;
+  }
```
This is the same as
```suggestion
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
```

alshakero commented on components/WhyMeditate/WhyMeditate.js:
```diff
@@ -0,0 +1,128 @@
+          </button>
+        )}
+
+        <audio
```
Let's preload the sound, so people don't wait when they click play.

```suggestion
        <audio preload
```

alshakero commented on components/WhyMeditate/AudioWithSubs.js:
```diff
@@ -0,0 +1,76 @@
+  });
+}
+
+export function AudioWithSubs({ time }) {
```
This should be called `Captions` now.
```suggestion
export function Captions({ time }) {
```

alshakero reviewed (changes requested):
Looks absolutely amazing. Tiny adjustments.

alshakerM committed:
Address feedback

alshakerM commented:
Changed the position of the speed button, the name of the component to Captions.
Added preload to audio

alshakero commented on components/WhyMeditate/Captions.module.css:
```diff
@@ -0,0 +1,45 @@
+}
+
+@keyframes appear {
+  0% {
+    color: #a85a5d;
+    border-radius: 10px;
+  }
+  20% {
+    opacity: 1;
+  }
+  100% {
+    opacity: 1;
+    background: transparent;
+  }
+}
```
The 0% changes the color and border radius, but the 100% changes the bg and opacity. They should agree on changing the same things.

alshakerM replied:
I have deleted the animation, since we are using SVG text.

alshakero commented on components/WhyMeditate/WhyMeditate.module.css:
```diff
@@ -0,0 +1,105 @@
+
+.playerContainer {
+  position: relative;
+  margin-left: 5%;
```
This margin is way too big. I think 2% is enough. Especially before playing. Also this needs a bit more margin top.

alshakero commented on components/WhyMeditate/WhyMeditate.module.css:
```diff
@@ -0,0 +1,105 @@
+  stroke: #a85a5d;
+  transform: rotate(270deg);
+  transform-origin: center center;
+  stroke-dasharray: 760;
+  stroke-dashoffset: 760;
+  stroke-dashoffset: calc(756 - (756 * var(--currentProgress)));
```
Extraneous dasharray.

```suggestion
  stroke-dasharray: 760;
  stroke-dashoffset: calc(756 - (756 * var(--currentProgress)));
```

alshakero commented on components/WhyMeditate/WhyMeditate.module.css:
```diff
@@ -0,0 +1,105 @@
+  stroke-dashoffset: calc(756 - (756 * var(--currentProgress)));
+}
+
+.audio {
+  margin-top: 10px;
+}
```
Does it need a margin when it's invisible?

alshakerM replied:
No, deleted

alshakero commented on components/WhyMeditate/WhyMeditate.module.css:
```diff
@@ -0,0 +1,105 @@
+.audio {
+  margin-top: 10px;
+}
+.progress {
+  stroke-linecap: round;
+}
```
Can we move this into progress above?

alshakero commented on components/WhyMeditate/WhyMeditate.module.css:
```diff
@@ -0,0 +1,105 @@
+  height: 48px;
+  padding: 0 4%;
+}
+.audioContainer {
+  background-color: #fffffffa;
+  border-radius: 50%;
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+  align-items: center;
+  position: absolute;
+  width: 100%;
+  height: 100%;
+  background-clip: content-box;
+  overflow: hidden;
+}
```
This is unused now.

```suggestion
.audioContainer {
  background-color: #fffffffa;
  border-radius: 50%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  position: absolute;
  width: 100%;
  height: 100%;
  background-clip: content-box;
  overflow: hidden;
}
```

alshakero commented on components/WhyMeditate/WhyMeditate.module.css:
```diff
@@ -0,0 +1,105 @@
+  overflow: hidden;
+}
+
+.audioContainer.isPlayed {
+  left: 0;
+  top: 0;
+
+  border: 15px solid rgba(0, 0, 0, 0);
+}
```
This is unused now.

alshakero commented on components/WhyMeditate/WhyMeditate.js:
```diff
@@ -0,0 +1,129 @@
+    }
+    audioRef.current.playbackRate = playbackRateNumbers[speedIndex];
+  }, [pause, speedIndex]);
+  const [time, setTime] = React.useState(0);
```
You can move this hook next to the others for better readability.

alshakero commented on components/WhyMeditate/Captions.module.css:
```diff
@@ -0,0 +1,45 @@
+  opacity: 0;
+}
+
+.specialWord {
+  fill: #a85a5d;
+  font-size: 40px;
+  opacity: 0;
+  position: absolute;
+  top: 40%;
+  animation: slideLeft 3.5s ease-in forwards;
+}
+@keyframes slideLeft {
+  50% {
+    opacity: 0.5;
+  }
+  100% {
+    opacity: 0;
+  }
+}
```
All this can be deleted now.

alshakero commented on components/WhyMeditate/WhyMeditate.js:
```diff
@@ -0,0 +1,129 @@
+          )}
+
+          {isPlayed && (
+            <>
+              <Captions time={time} />
+            </>
```
Extraneous fragment.

alshakerM replied:
Why, if I remove it we get no text with our audio

alshakero replied:
I mean the fragment only `<></>` you have one child anyways.

alshakero commented on components/WhyMeditate/WhyMeditate.js:
```diff
@@ -0,0 +1,129 @@
+            setTime(e.currentTarget.currentTime);
+          }}
+          src="/sound.mp3"
+          style={{ width: 500 }}
```
Unused styles.
```suggestion
          
```

alshakero commented on components/WhyMeditate/WhyMeditate.module.css:
```diff
@@ -0,0 +1,105 @@
+  font-style: italic;
+  color: #a85a5d;
+  position: relative;
+  animation: delay 1s ease-out;
+}
+@keyframes delay {
+  0% {
+    opacity: 0;
+  }
+
+  100% {
+    opacity: 1;
+  }
+}
```
I don't think this is doing anything.

alshakerM replied:
It is, but only on play not pause

alshakero reviewed (changes requested):
Left a few janitorial comments. There are some leftovers from all the iterations.


Also, please address the conflicts.

alshakerM committed:
Merge branch 'master' into add-why-meditate-page

alshakerM committed:
Address feedback

alshakerM committed:
Move the time hook back with the rest of the hooks

alshakero reviewed (approved):
Looks mejestic. LGTM!
